src/VisualStudio/Core/Def/Workspace/VisualStudioSourceGeneratorTelemetryCollectorWorkspaceServiceFactory.cs

```diff
@@ -0,0 +1,98 @@
+            // workspaces (interactive, for example) will be put in a different bucket. This allows us to report the telemetry
+            // from the primary workspace on solution closed, while not letting the unrelated runs pollute those numbers.
+
+            // The expectation is there aren't really generators running in other workspaces, so that data is likely empty.
+            if (workspaceServices.Workspace is VisualStudioWorkspace)
+            {
+                EnsureSubscribedToSolutionEvents();
+                return _visualStudioWorkspaceInstance;
+            }
+            else
+            {
+                return _otherWorkspacesInstance;
+            }
+        }
+
+        private void EnsureSubscribedToSolutionEvents()
+        {
+            if (Interlocked.CompareExchange(ref _subscribedToSolutionEvents, 1, 0) == 0)
+            {
+                Task.Run(async () =>
+                {
+                    var shellService = await _serviceProvider.GetServiceAsync<SVsSolution, IVsSolution>(_threadingContext.JoinableTaskFactory).ConfigureAwait(true);
+                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(_threadingContext.DisposalToken);
+                    shellService.AdviseSolutionEvents(this, out _);
+                }, _threadingContext.DisposalToken);
+            }
+        }
+
+        public void ReportOtherWorkspaceTelemetry()
+        {
+            _otherWorkspacesInstance.ReportStatisticsAndClear(FunctionId.SourceGenerator_OtherWorkspaceSessionStatistics);
+        }
+
+        int IVsSolutionEvents.OnAfterOpenProject(IVsHierarchy pHierarchy, int fAdded) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnQueryCloseProject(IVsHierarchy pHierarchy, int fRemoving, ref int pfCancel) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnBeforeCloseProject(IVsHierarchy pHierarchy, int fRemoved) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnAfterLoadProject(IVsHierarchy pStubHierarchy, IVsHierarchy pRealHierarchy) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnQueryUnloadProject(IVsHierarchy pRealHierarchy, ref int pfCancel) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnBeforeUnloadProject(IVsHierarchy pRealHierarchy, IVsHierarchy pStubHierarchy) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnAfterOpenSolution(object pUnkReserved, int fNewSolution) => VSConstants.E_NOTIMPL;
+        int IVsSolutionEvents.OnQueryCloseSolution(object pUnkReserved, ref int pfCancel) => VSConstants.E_NOTIMPL;
+
+        int IVsSolutionEvents.OnBeforeCloseSolution(object pUnkReserved)
+        {
+            // Report the telemetry now before the solution is closed; since this will be reported per solution session ID, it means
+            // we can distinguish how many solutions have generators versus just overall sessions.
```

Comment:
If there's a different way to do this by capturing the context when the solution is opened or something else, I'd love to know it.